### PR TITLE
skip running stages

### PR DESCRIPTION
**Pull request by eesmerdag**

### CODE REVIEW

**Comment by eesmerdag** [comment]
https://www.notion.so/figmentnetworks/Find-a-best-way-to-speed-up-the-process-of-backfilling-data-354c1ef5a2c84b708d281159d0d29393

File: go.sum [hunk 1]
```diff
@@ -1,12 +1,48 @@
+github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
```
Comment (pacmessica) [inline go.sum:1]: lol wow this was out of date

File: pipeline/pipeline_test.go [hunk 2]
```diff
@@ -114,7 +119,7 @@ func TestPipeline_SetStages(t *testing.T) {
 
 		options := &pipeline.Options{}
 
-		if err := p.Start(ctx, &sourceMock{1, 2, 1}, sinkMock, options); err != nil {
+		if err := p.Start(ctx, &sourceMock{1, 2, 1, false}, sinkMock, options); err != nil {
```
Comment (pacmessica) [inline pipeline/pipeline_test.go:122]: can you add a new test for when  `skipRunningStagesForCurrentHeight` = true?

**Review by pacmessica — COMMENTED** [review]
looks good! Just add a test and update the README and then it'll look extra good 😄

File: example/custom/indexing/source.go [hunk 3]
```diff
@@ -14,10 +14,10 @@ func NewSource() pipeline.Source {
 }
 
 type source struct {
-	startHeight   int64
-	currentHeight int64
-	endHeight     int64
-	err           error
+	startHeight                       int64
```
Comment (silentlight) [inline example/custom/indexing/source.go:17]: I think these spaces are too much :p

File: example/default/indexing/source.go [hunk 4]
```diff
@@ -14,10 +14,10 @@ func NewSource() pipeline.Source {
 }
 
 type source struct {
-	startHeight   int64
-	currentHeight int64
-	endHeight     int64
-	err           error
+	startHeight                       int64
```
Comment (silentlight) [inline example/default/indexing/source.go:17]: Same here

File: example/custom/indexing/source.go [hunk 5]
```diff
@@ -35,3 +35,7 @@ func (s *source) Current() int64 {
 func (s *source) Err() error {
 	return s.err
 }
+
+func (s *source) SkipRunningStagesForCurrentHeight() bool {
```
Comment (silentlight) [inline example/custom/indexing/source.go:39]: I think just `Skip` or `ShouldSkip` will be enough here since other method names are really concise.

**Review by silentlight — COMMENTED** [review]
Good job on unit tests. Left few minor comments.

File: example/custom/indexing/source.go (hunk 3 above)
Comment (eesmerdag) [reply example/custom/indexing/source.go:17]: 👍 😢

File: example/custom/indexing/source.go (hunk 5 above)
Comment (eesmerdag) [reply example/custom/indexing/source.go:39]: 👍

File: pipeline/pipeline_test.go [hunk 6]
```diff
@@ -193,7 +194,67 @@ func TestPipeline_Start(t *testing.T) {
...
+		runFetcher.After(runSyncer)
+		runParser.After(runFetcher)
+		runValidator.After(runParser)
+
+		runSequencer.After(runValidator)
+		runAggregator.After(runValidator)
+
+		runPersistor.After(runSequencer)
+		runPersistor.After(runAggregator)
+		runCleanup.After(runPersistor)
```
Comment (pacmessica) [inline pipeline/pipeline_test.go:252]: Don't need these `.After` statements since you don't expect them to run

Comment (eesmerdag) [reply pipeline/pipeline_test.go]: 👍

File: pipeline/interfaces.go [hunk 7]
```diff
@@ -34,7 +34,7 @@ type Source interface {
 // Sink is executed as a last stage in the pipeline
 type Sink interface {
 	// Consume consumes payloadMock
-	Consume(context.Context, Payload, bool) error
```
Comment (pacmessica) [inline pipeline/interfaces.go]: ahh I forgot sink.Consume was an interface. nice

**Review by pacmessica — APPROVED** [review]

**Review by silentlight — APPROVED** [review]
Looks good @eesmerdag. Nice work!

**Review by silentlight — COMMENTED** [review]
Let's talk about it today cause I think that we can make it even more flexible and avoid using those if statements in the sink

**Pull request merged**